File: claude/src/fetchers/openalex.mjs
import axios from 'axios';
import { isTitleMatch } from '../utils/titleMatch.mjs';

/**
 * OpenAlex fetcher - free and open catalog of scholarly papers
 * Now validates title similarity to avoid false positives
 */
export async function fetchFromOpenAlex(title) {
  try {
    const searchUrl = 'https://api.openalex.org/works';
    const params = {
      search: title,
      per_page: 5,
      mailto: process.env.UNPAYWALL_EMAIL || 'example@example.com'
    };

    const response = await axios.get(searchUrl, { params, timeout: 10000 });

    if (!response.data.results || response.data.results.length === 0) {
      return { success: false, error: 'No results found on OpenAlex' };
    }

    // Try to find a paper with matching title AND open access PDF
    for (const work of response.data.results) {
      // Validate title similarity first
      if (!isTitleMatch(title, work.title)) {
        continue;  // Skip non-matching titles
      }

      const pdfUrl = work.open_access?.oa_url ||
                     work.primary_location?.pdf_url ||
                     work.best_oa_location?.pdf_url;

      if (pdfUrl) {
        return {
          success: true,
          pdf_url: pdfUrl,
          source: 'OpenAlex',
          metadata: {
            title: work.title,
            authors: work.authorships?.map(a => a.author?.display_name).filter(Boolean).join(', '),
            year: work.publication_year,
            doi: work.doi?.replace('https://doi.org/', ''),
            isOpenAccess: work.open_access?.is_oa
          }
        };
      }
    }

    // Also return DOI if found for matching title (for Unpaywall fallback)
    for (const work of response.data.results) {
      if (isTitleMatch(title, work.title) && work.doi) {
        return { 
          success: false, 
          error: 'No open access PDF found on OpenAlex',
          doi: work.doi.replace('https://doi.org/', '')
        };
      }
    }

    return { success: false, error: 'No matching paper found on OpenAlex' };
  } catch (error) {
    return { success: false, error: `OpenAlex error: ${error.message}` };
  }
}

File: codex/app/sources/openalex.py
from app.sources.base import SourceContext
from app.utils import is_title_match, safe_title_query


def _extract_pdf(data: dict) -> str | None:
    oa = data.get("open_access") or {}
    if oa.get("oa_url"):
        return oa.get("oa_url")
    primary = data.get("primary_location") or {}
    if primary.get("pdf_url"):
        return primary.get("pdf_url")
    return None


async def fetch_openalex(ctx: SourceContext, doi: str | None, title: str | None) -> dict | None:
    if doi:
        url = f"https://api.openalex.org/works/https://doi.org/{doi}"
        data = await ctx.get_json(url)
        pdf_url = _extract_pdf(data)
        if not pdf_url:
            doi_value = data.get("doi")
            if doi_value:
                return {
                    "pdf_url": None,
                    "doi": doi_value,
                    "metadata": data,
                    "source": "openalex",
                }
            return None
        return {"pdf_url": pdf_url, "metadata": data, "source": "openalex"}

    if not title:
        return None

    url = f"https://api.openalex.org/works?search={safe_title_query(title)}&per-page=1"
    data = await ctx.get_json(url)
    results = data.get("results") or []
    if not results:
        return None
    hit = results[0]
    if title and not is_title_match(title, hit.get("title", "")):
        return None
    pdf_url = _extract_pdf(hit)
    if not pdf_url:
        doi_value = hit.get("doi")
        if doi_value:
            return {
                "pdf_url": None,
                "doi": doi_value,
                "metadata": hit,
                "source": "openalex",
            }
        return None
    return {"pdf_url": pdf_url, "metadata": hit, "source": "openalex"}
